src/txn/txn.c

```diff
@@ -1484,7 +1484,10 @@ __wt_txn_release_resources(WT_SESSION_IMPL *session)
 {
     WT_TXN *txn;
 
-    txn = &session->txn;
+    if ((txn = session->txn) == NULL)
+        return;
+
+    __wt_txn_clear_timestamp_queues(session);
```

Comment:
This surprises me a little - why is it here?

Reply:
@agorrod 
> This surprises me a little - why is it here?

Currently, we remove the `WT_TXN` from the timestamp queues when we clear the `WT_SESSION` structure, so in `__session_clear()`, which is called at `WT_SESSION.close()`.

(Originally there was just a commit queue, and later we added timestamp queues. Currently it's the `durable_timestamp` and `read_timestamp` queues.)

That looked wrong to me, so I switched the code to do remove the `WT_TXN` from the queues when we release transaction resources, that is, in `__wt_txn_release_resources()`, so it happens whenever we wrap up a transaction.

I should probably have thought harder about that change: this moves the work from being rare, to being relatively frequent and there are locks involved in messing with the timestamp queues, so maybe the original decision was deliberate. It seems weird to me that we're hanging out on the timestamp queues even when we're no longer in a transaction, but, maybe? And, of course, the queues we're dealing with have changed since then, so it's probably worth a check regardless.

@sueloverso, @michaelcahill: Sue, this was originally your change, do you remember why we pull the `WT_TXN` structure off of the timestamp queues when we close the session, rather than when we release the transaction from further work?

Reply:
@keithbostic if you look at the ticket for that change originally, it was specifically to not acquire a write lock on committing a transaction. Cutting the number of write locks in half had a large performance impact. (Or really regained lost performance when we were acquiring it.) So that change was deliberate that we set the `clear` field and leave it on the queue until a thread that already has the lock anyway can do maintenance.

Reply:
https://jira.mongodb.org/browse/WT-3854

Reply:
Thanks, @sueloverso.

That's a problem for this branch, because the timestamp queues currently store `WT_TXN` references and this change would mean those references point into free'd memory once the transaction completes.

I'm going to see what @tetsuo-cpp's PR #5485 ends up looking like and then get back to this branch. My suspicion is @tetsuo-cpp will move the timestamp queue fields from the `WT_TXN` structure into the `WT_TXN_SHARED/WT_TXN_STATE` structure (which makes sense), and that means this branch will work. If that's not the case, I'll just kill this branch.

Reply:
> That looked wrong to me, so I switched the code to do remove the WT_TXN from the queues when we release transaction resources, that is, in __wt_txn_release_resources(), so it happens whenever we wrap up a transaction.

Ok, perhaps I'm interpreting the phrase 'wrap up a transaction' differently than intended. I thought you meant on every commit or rollback. Tracing the code back, I see `wt_txn_release_resources` is only called from two locations:
#. `wt_txn_destroy`: This is only called from `session_close` so this is not on a critical path.
#. `wt_session_release_resources`: This is called from the following:
* `session_close`: non-critical path
* `session_reset`: non-critical path
* `session_checkpoint`: not sure we want checkpoint needing to acquire two more write locks.
* `wt_session_compact`: non-critical path

I am relieved it is not on commit and rollback, however, I don't see benefit in this part of the change and possible detriment from the checkpoint call. I'd suggest reverting this part.

Reply:
Yes, it's not on commit/abort, it's only when the session releases its resources in general. I hadn't actually worked through the call stack, thanks for that, I was assuming it happened on more "interesting" code paths.

I'm not worried about checkpoint, honestly. That's a heavy-weight operation, 2 locks won't matter, once a minute.

Reply:
@keithbostic I can confirm that I've gone ahead and moved the timestamp queue information to `WT_TXN_SHARED`.

Reply:
I don't feel strongly about this change overall.

I do think we should be careful about `WT_SESSION::reset`: MongoDB calls that every time one of its operations completes (before putting a session back in the cache).  So maybe kick off a patch build of some MongoDB perf tests before committing to this part of the change?

Reply:
Understood about `WT_SESSION.reset`.

My expectation is that once @tetsuo-cpp merges his changes this can go back to the way it was before I moved that call, so absent a reason to move it, performance should be identical. If that's not the case, I'll either test or kill the branch.